src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs

```diff
@@ -283,6 +285,79 @@ public class LockFileBuilder
             return lockFile;
         }
 
+        private void ValidateLockFileLibrary(IList<LockFileLibrary> libraries)
+        {
+            var libNameVersion = libraries.Select(l => l.Name + " " + l.Version);
```

Comment:
Would it be at all simpler to just create a dictionary, and iterate over the libs one by one and updating the selected ref as we go?

Currently

* Distinct uses a set internally - 1 pass: https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Linq/src/System/Linq/Distinct.cs. 

* The ToLookup statement creates a dict and then it iterates over that, allocating a list there. 
* Finally, iterate over the main list to remove the dups. 

All of that can be replaced with 1 dictionary and 1 loop. 
The moment you find a conflict, rank references and dedup. 

Same applies to both methods.

Reply:
@nkolev92 
Could you be able to check now? I pushed new changes.